enterprise/server/remote_execution/dirtools/dirtools.go

```diff
@@ -702,67 +751,76 @@ func (ff *BatchFileFetcher) FetchFiles(filesToFetch FileMap, opts *DownloadTreeO
 			continue
 		}
 
-		// Attempt to link files from the local file cache.
-		numFilesLinked := 0
-		for _, fp := range filePointers {
-			d := fp.FileNode.GetDigest()
-			if fileCache != nil {
-				linked, err := linkFileFromFileCache(ff.ctx, d, fp, fileCache, opts)
-				if err != nil {
-					return err
-				}
-				if !linked {
-					break
-				}
-				numFilesLinked++
+		linkEG.Go(func() error {
+			// If we linked the digest from the file cache, there's nothing
+			// more to do.
+			if err := ff.linkFromFileCache(filePointers, opts); err == nil {
+				return nil
 			}
-		}
-		// If we successfully linked all files from the file cache, no need to
-		// download this digest. Note: We may not successfully link all files
-		// if the digest expires from the cache while the above loop is in progress.
-		if numFilesLinked == len(filePointers) {
-			continue
-		}
 
-		// At this point we need to download the contents of the digest.
-		// If the file exceeds our gRPC max size, it'll never
-		// fit in the batch call, so we'll have to bytestream
-		// it.
-		size := d.GetSizeBytes()
-		if size > gRPCMaxSize || ff.env.GetContentAddressableStorageClient() == nil {
-			func(d *repb.Digest, fps []*FilePointer) {
+			// Otherwise, queue the digest to be fetched.
+			fetchQueue <- digestToFetch{d: d, fps: filePointers}
+			return nil
+		})
+	}
+
+	// Read work off the fetchQueue channel and generate batch read requests
+	// to download data.
+	eg, ctx := errgroup.WithContext(ff.ctx)
+	enqueueWorkDone := make(chan struct{})
+	go func() {
```

Comment:
did you consider using the errgroup for this goroutine as well? That might let you get rid of enqueueWorkDone and just wait on the eg